### DISCUSSION TITLE
In memory syncing between ballerina replications

### DISCUSSION THREAD

**Discussion by pcnfernando** [ballerina-platform/ballerina-lang#39510] [Ideas]
Assume, I have a stateful ballerina program(uses in-memory storage through tables, arrays, maps,  etc).
When this program is deployed in Choreo and needs to guarantee availability; when we do replications on the app we'll have to migrate the original ballerina source from in-memory to DB-backed or other means.

I was wondering instead of a source change, we provide means through the language level to sync the in-memory between the instances.

Suggestion from the language level,
-  Annotate the in-memory constructs to use an in-memory syncing logic (eg:redis)

**Comment by pcnfernando** [comment]
Looping @hasithaa @gimantha. Appreciate your thoughts on this

**Reply by hasithaa** [reply]
Are you looking for an entity management layer? I think we should be able to extend new persistence design to redis. @daneshk thoughts.

**Reply by pcnfernando** [reply]
Yes, something like an entity management layer. Extending the new persistence design to redis/hazelcast should work I guess.

Just to give a bit more context.
To scale a stateful application there are several ways and we can even use a combination of these ways. For the moment let's ignore vertical scaling, caching, and other scenarios where we don't need multiple instances of an application and consider horizontal scaling, sharding, replication, load balancing, and so on. 
For each of these cases, we need a syncing mechanism for the data. I was considering in-memory syncing AFIK which is better for low latency, high throughput data access, and having a relatively small amount of data that can fit in memory compared to having a messaging layer or DB access, etc;
